charmbits/httpservice/server.go

```diff
@@ -0,0 +1,439 @@
+	c, err := startServer0(ctxt, args, h)
+	if err != nil {
+		return c, err
+	}
+	return c, err
+}
+
+// start starts the actual HTTP server running. It is called
+// in service context, not hook context.
+func startServer0(ctxt *service.Context, args []string, h *handlerInfo) (hook.Command, error) {
+	if len(args) != 1 {
+		return nil, fmt.Errorf("need exactly one argument, got %q", args)
+	}
+	srv := &server{
+		handlerInfo: h,
+		stateDir:    args[0],
+	}
+	state, err := srv.loadState()
+	if err != nil {
+		return nil, fmt.Errorf("cannot load state: %v", err)
+	}
+	var fb Feedback
+	srv.set(state, &fb)
+	fb.show()
+	// Fool net/rpc into thinking the type is exported, because
+	// we don't want to export the srv type from this package.
+	type Srv struct {
+		*server
+	}
+	rpcCmd, err := ctxt.ServeLocalRPC(Srv{srv})
+	if err != nil {
+		return nil, errgo.Notef(err, "serve local RPC failed")
+	}
+	srv.tomb.Go(func() error {
+		srv.tomb.Kill(rpcCmd.Wait())
+		return nil
+	})
+	srv.tomb.Go(func() error {
+		<-srv.tomb.Dying()
+		// The command has been killed. Stop the RPC listener
+		// and close any current listeners to cause the handlers
+		// to terminate.
+		rpcCmd.Kill()
+		err := rpcCmd.Wait()
+		if err != nil {
+			err = errgo.Notef(err, "local RPC server")
+		}
+
+		srv.mu.Lock()
+		defer srv.mu.Unlock()
+		srv.closeResources()
+		return err
+	})
+	return srv, nil
+}
+
+// Kill implements gocharm.Command.Kill.
+func (srv *server) Kill() {
+	srv.tomb.Kill(nil)
+}
+
+// Wait implements gocharm.Command.Wait.
+func (srv *server) Wait() error {
+	return srv.tomb.Wait()
+}
+
+// Feedback is an implementation detail, exposed only because net/rpc requires it.
+type Feedback struct {
+	Warnings []string
+}
+
+// show shows all the warnings.
+func (fb *Feedback) show() {
+	for _, w := range fb.Warnings {
+		fmt.Fprintf(os.Stderr, "warning: %v\n", w)
+	}
+}
+
+// addError adds the given error to the list of warnings
+// that will be reported as the result of an RPC call.
+func (fb *Feedback) addError(err error) {
+	fb.Warnings = append(fb.Warnings, err.Error())
+}
+
+// Changed implements an RPC server method that is called
+// when a value changes in the charm.
+func (srv *server) Set(p *ServerState, response *Feedback) error {
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+	srv.set(*p, response)
+	srv.state = *p
+	// Save the server state so that if it's started or the
+	// instance is rebooted, we can carry on as before.
+	if err := srv.saveState(srv.state); err != nil {
+		return errgo.Notef(err, "cannot save server state")
+	}
+	return nil
+}
+
+func (srv *server) serveHTTP(port int, h http.Handler) error {
+	addr := ":" + strconv.Itoa(port)
+	listener, err := net.Listen("tcp", addr)
+	if err != nil {
+		return errgo.Newf("cannot listen on %s: %v", addr, err)
+	}
+	server := &http.Server{
+		Addr:    addr,
+		Handler: h,
+	}
+	srv.httpListener = newHandlerListener(server, listener)
```

Comment:
newHandlerListener can't fail. I've changed newHandlerListener to return the *handlerListener instead of setting the field, as that makes the logic a little more obvious inside the set method.